retrieval/discovery/ec2.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/log"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	ec2NameLabel = model.MetaLabelPrefix + "ec2_instance_id"
+)
+
+// EC2Discovery periodically performs EC2-SD requests. It implements
+// the TargetProvider interface.
+type EC2Discovery struct {
+	aws            *aws.Config
+	tags           model.LabelSet
+	regexTags      map[string]*regexp.Regexp
+	done           chan struct{}
+	interval       time.Duration
+	port           int
+	public         bool
+	instanceFilter []*ec2.Filter
+}
+
+// NewEC2Discovery returns a new EC2Discovery which periodically refreshes its targets.
+func NewEC2Discovery(conf *config.EC2SDConfig) *EC2Discovery {
+	creds := credentials.NewStaticCredentials(conf.Access, conf.Secret, "")
+	if conf.Access == "" && conf.Secret == "" {
+		creds = credentials.NewEnvCredentials()
+	}
+	tags := make(model.LabelSet)
+	filter := &ec2.Filter{
+		Name: aws.String("tag-key"),
+	}
+	for k, v := range conf.Tags {
+		tags[model.LabelName(k)] = model.LabelValue(v)
+		filter.Values = append(filter.Values, &k)
+	}
+	return &EC2Discovery{
+		aws: &aws.Config{
+			Region:      &conf.Region,
+			Credentials: creds,
+		},
+		tags:      tags,
+		regexTags: conf.RegexTags,
+		done:      make(chan struct{}),
+		interval:  time.Duration(conf.RefreshInterval),
+		port:      conf.Port,
+		public:    conf.Public,
+		instanceFilter: []*ec2.Filter{
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running")},
```

Comment:
No, I'm not sure. I can remove this. Thought it was an easy performance win since it limited the instances being returned to useable ones. I think the tag-key filter is good enough for that, though.